Implement Serialization and Deserialization for Verbosity

joshka opened the pull request:
Verbosity is serialized and deserialized using the lowercase of the
VerbosityFilter (e.g. "debug")

The `serde` dependency is gated behind an optional feature flag.

Added conversion methods between Verbosity and VerbosityFilter to
simplify the implementation and derived PartialEq, Eq impls on
types where this was necesary for testing.

Fixes: https://github.com/clap-rs/clap-verbosity-flag/issues/88

epage commented on src/lib.rs:
```diff
@@ -142,19 +157,17 @@ fn level_value(level: Option<Level>) -> i8 {
 }
 
-fn level_enum(verbosity: i8) -> Option<Level> {
+fn level_enum(verbosity: i16) -> Option<Level> {
```
> The internal representation in calculations has been changed from i8 to
i16 to prevent overflows when performing calculations using u8 values.


Please make this its own commit

joshka replied:
#115

epage commented on src/lib.rs:
```diff
@@ -126,12 +131,22 @@ impl<L: LogLevel> Verbosity<L> {
-    fn verbosity(&self) -> i8 {
-        level_value(L::default()) - (self.quiet as i8) + (self.verbose as i8)
+    fn verbosity(&self) -> i16 {
+        level_value(L::default()) - (self.quiet as i16) + (self.verbose as i16)
```
How bad would it be if we instead switched this to
```suggestion
        level_value(L::default()).saturating_add((self.verbose - self.quiet)))
```

epage replied:
In general, people shouldn't be providing so many that they saturate

joshka replied:
A single `-q` would underflow (0u8 - 1u8). This would work fine in debug mode, and fail in release mode.

While in general that is indeed true that users calling the executable should not in general write paramaters that would cause it to crash, similarly apps should be robust in the face of users doing strange things. I don't think it's safe for a library to crash with an overflow given bad inputs like this. It may be better to clamp the total number of parameters parsed to some given number.

Given Clap is used to process command lines, that some commands will be part of administrative scripts running with elevated privileges, being able to control a crash in a command with specific parameters lead to potential (very low impact) CVEs.

It's possible that switching this internally to u8 would be a better idea, and make the OFF condition return 0 instead of -1. This would allow for u8 and saturating to 0 throughout (except for when calculating the delta between quote and verbosity).

I agree it makes sense to pull this change out as separate commit. I think it probably makes sense as a new PR for review purposes, or is the idea of this split to pull out this change strictly for commit history / message purposes?

joshka replied:
Switched to u8 instead in #115

epage commented on src/serde.rs:
```diff
@@ -0,0 +1,336 @@
+                formatter.write_str("a log level")
+            }
+
+            fn visit_none<E: de::Error>(self) -> Result<Self::Value, E> {
+                let verbose = 0;
+                let quiet = (level_value(L::default()) + 1) as u8;
+                Ok(Verbosity::new(verbose, quiet))
+            }
+
+            fn visit_some<D: Deserializer<'de>>(
+                self,
+                deserializer: D,
+            ) -> Result<Self::Value, D::Error> {
+                let level = Level::deserialize(deserializer)?;
+                let value = level_value(Some(level)) - level_value(L::default());
+                let verbose = value.max(0) as u8;
+                let quiet = (-value).max(0) as u8;
+                Ok(Verbosity::new(verbose, quiet))
+            }
+        }
```
Please split this off into its own private construcor for taking an `Option<Level>` with tests for that

joshka replied:
No longer necessary in 2dda2ab

epage commented on Cargo.toml:
```diff
@@ -112,16 +112,22 @@ pre-release-replacements = [
 [dependencies]
 log = "0.4.1"
 clap = { version = "4.0.0", default-features = false, features = ["std", "derive"] }
+serde = { version = "1.0", optional = true, features = ["derive"] }
```
Did I miss seeing a `derive` being used?

joshka replied:
No, I included this and then realized this required manual serialization. Removing.

joshka replied:
With the simplification in 2dda2ab to use the `From` implemenation instead of a custom implementation, I added the derive feature back in.

epage commented on Cargo.toml:
```diff
@@ -112,16 +112,22 @@ pre-release-replacements = [
 tracing-subscriber = "0.3"
 tracing-log = "0.2"
+serde_test = "1.0.177"
+rstest = "0.22.0"
```
I'd rather not use this at this time

joshka replied:
Ok, I'll remove rstest. Do you have a preference for writing tests that only vary in their parameterization to replace rstest with? E.g. internal closure, helper functions, repeated code, assertions with test specific messages, some other way?

For curiousity's sake, what's the rationale behind avoiding rstest?

joshka replied:
Removed in 2dda2ab

epage commented on src/serde.rs:
```diff
@@ -0,0 +1,336 @@
+
+            fn visit_none<E: de::Error>(self) -> Result<Self::Value, E> {
+                let verbose = 0;
+                let quiet = (level_value(L::default()) + 1) as u8;
+                Ok(Verbosity::new(verbose, quiet))
```
Hmm, should we be serializing levels of level filters?

We should specifically test for how this interacts with `toml` as that doesn't have the concept of `None`

joshka replied:
I think `LevelFilter` would be better than `Level`. Serializing OFF/ERROR/... instead of None/Some(ERROR)/... makes more sense. I went with the conservative approach (`Level`) more for an assumed consistency than because I thought it was the right approach. Will change to `LevelFilter`.

joshka replied:
Switched to `LevelFilter` in 2dda2ab. Adding `impl From<LevelFilter>` and the reverse made it easy to derive serialization instead of this being manual.

joshka reviewed (commented):
I will:
- remove rstest
- serialize to/from LevelFilter instead of Option(Level)
- add a specific test for TOML serialization
- Fix the github lints
- split the i8 change out to another commit (and will raise another PR for it if necessary)

joshka reviewed (commented):
Moving to LevelFilter made it easy to derive serialize / deserialize using `From<LevelFilter>` instead of a custom implementation.

joshka commented on src/lib.rs:
```diff
@@ -163,8 +176,24 @@ impl<L: LogLevel> fmt::Display for Verbosity<L> {
     }
 }
 
+impl<L: LogLevel> From<Verbosity<L>> for LevelFilter {
+    fn from(v: Verbosity<L>) -> Self {
+        v.log_level_filter()
+    }
+}
+
+impl<L: LogLevel> From<LevelFilter> for Verbosity<L> {
+    fn from(filter: LevelFilter) -> Self {
+        let default_verbosity = level_value(L::default());
+        let verbosity = level_filter_value(filter);
+        let verbose = verbosity.saturating_sub(default_verbosity);
+        let quiet = default_verbosity.saturating_sub(verbosity);
+        Verbosity::new(verbose, quiet)
+    }
+}
+
 /// Customize the default log-level and associated help
-pub trait LogLevel {
+pub trait LogLevel: Clone {
```
Clone is required because of the `#[serde(into = "LevelFilter")]` setting (https://serde.rs/container-attrs.html#into).

This is technically a semver breaking change, as now implementors of LogLevel need to make their implementation implement Clone. I think there may be a way to fix this with another serde attribute.

joshka replied:
`#[serde(bounds(serialize = "L: Clone")))` does the trick.

joshka commented on src/lib.rs:
```diff
@@ -63,8 +63,10 @@ pub use log::Level;
 
 /// Logging flags to `#[command(flatten)]` into your CLI
-#[derive(clap::Args, Debug, Clone, Default)]
+#[derive(clap::Args, Debug, Clone, Default, PartialEq, Eq)]
```
Added to support test assertions

epage commented on src/lib.rs:
```diff
@@ -126,30 +126,32 @@ impl<L: LogLevel> Verbosity<L> {
 
-    fn verbosity(&self) -> i8 {
-        level_value(L::default()) - (self.quiet as i8) + (self.verbose as i8)
+    fn verbosity(&self) -> u8 {
```
Looks like this PR is blocked on #115

joshka replied:
Yep, I split off the changes to make the u8 parts of the change easier to evaluate on their own. Github doesn't have a good way to do stacked PRs unfortunately.

epage replied:
I know, I deal with that a lot.  This is more of a bookkeeping comment to make sure we don't move forward without that being merged first.  When contributing, I recommend making these relationships explicit, rather than relying on maintainers to catch them.

epage replied:
At this point, this PR needs a rebase and this item can be resolved.

epage commented on src/lib.rs:
```diff
@@ -396,4 +432,242 @@ mod test {
+    #[test]
+    fn from_level_filter() {
+        // ErrorLevel
+        let v = Verbosity::<ErrorLevel>::from(LevelFilter::Off);
+        assert_eq!(v, Verbosity::new(0, 1));
```
I think more important than what quiet/verbose levels are is to see that it produces the same `LevelFiltter`

btw for these tests, you can always iterate over those of the same T in a loop, with an array of inputs and expecteds.

joshka replied:
Subjectively (based on my experience), adding loops or conditionals to tests leads to worse tests. Tests in that form are as a general rule are more difficult to diagnose for failure. Often tests written this way don't take into account the main goal of testing, that tests should be useful when they fail. Specifically taking this approach here leads to tests that you have to interpret how the test is constructed and then which of the constructions led to failure. In one off situations that's fine, but as you add more and more of these types of tests then these little paper cuts build up. By taking this approach you also don't get to see all the failures, only the first failure in a potential list of failures. This makes fixing any errors when failure happens an O(N) rather than O(1) type problem.

I'm fine with making this change, but at the same time, I would ask whether it's worth your time to reconsider your position on using a tool like rstest for testing things which are naturally parameterized like this. You get the same result as a looped parameterized condition, but the complexity cost is amortized across all tests rather than paid by each test. I know that this may come across as belaboring a point (and apologize if it does come across as such). However the points you're bringing up about testing approach are solved problems with an existing tool, which when avoided means re-solving those same challenges manually. I also don't know what tradeoffs your considering with respect to not using rstest, so as mentioned I'm fine going with this approach (note the changes in https://github.com/clap-rs/clap-verbosity-flag/pull/115 at https://github.com/clap-rs/clap-verbosity-flag/pull/115/commits/7f435cde67459fc9a8c95c8bebd7a9808db39e0d

epage replied:
I know very well the value of parameterized tests.  I deal with enough custom test frameworks / harnesses that I do not want to deal with any more at this time.  They are all a hack on what should be provided by the de facto test harness which is what I'm working to improve as part of my efforts with the T-testind-devex team in the Rust Project.

joshka replied:
Got it - makes sense. And agree definitely on should be part of the defaults :)

epage commented on src/lib.rs:
```diff
@@ -396,4 +432,242 @@ mod test {
+        }
+
+        // round-trips
+        let toml = "meaning_of_life = 42\nverbose = \"DEBUG\"\n";
```
Not sure how I feel about deserializing to all-caps

joshka replied:
Yeah, this irks me a little too, but being consistent with the existing behavior won out over my ideal approach. If I was designing the `log` crate, I think I'd have made that serialization lower case too.

Here, this can deserialize from either case, but serializes to upper case. It comes from the serialization of `log::LevelFilter`. That means upper case is consistent with any other serialization that works with log level serialization (and is consistent with the usual display of the level in log outputs).

If you did want to avoid this, it would require going back to using a custom serialization implementation. This doesn't seem worth the extra code and tests to me. I prefer the simplicity even if that comes with a slightly ugly default which is consistent with other things doing the same thing.

I also suspect the serialization side of this will be used much less frequently than the deserialization side, and the serialization side will also often be more focused on programs re-reading the format rather than humans.

epage replied:
On one hand, *if* people are serializing `LevelFilter`, this keeps us consistent with that.  On the other hand, this will produce results inconsistent with every other serialized value.  This isn't just important here but for other cases like schemars.  I think I want to let this sit for a time to mull over this.

joshka replied:
no problem :)

joshka replied:
eb58de2d699758a72b437a83059d01c9179aa407 serializes to `VerbosityFilter` instead of this, which doesn't have a deserialization implementation that handles upper case tokens.

epage replied:
That replaced SCREAMING CASE with Title case
```rust
assert_eq!(toml, "meaning_of_life = 42\nverbosity = \"Debug\"\n");
```
I feel like lower case is much more likely to align with a users configuration format design (as most are likely kebab-case or snake_case)

epage replied:
btw eb58de2  just changed the value but it doesn't answer the question about interoperability

joshka commented:
In eb58de2d699758a72b437a83059d01c9179aa407, I rewrote this based on the 3.0 release. Notably this uses the `VerbosityFilter` to de/serialize instead of `log::LogLevelFilter` which means that it only handles `Title` case deserialization. I've also simplified the style and quantity of tests significantly.

joshka commented on src/lib.rs:
```diff
@@ -453,4 +488,84 @@
+
+#[cfg(feature = "serde")]
+mod serde_tests {
+    use super::*;
```
This is triggered because the exception for this is in modules named "tests", not ending in "tests". Added an `#[allow]`.

epage replied:
Looks like its actually from missing `#[cfg(test)]`.  Added it and I'm not seeing a lint.

epage commented on src/lib.rs:
```diff
@@ -70,8 +70,21 @@ pub mod log;
 
 /// Logging flags to `#[command(flatten)]` into your CLI
-#[derive(clap::Args, Debug, Clone, Default)]
+#[derive(clap::Args, Debug, Clone, Default, PartialEq, Eq)]
```
This PR seems to be mixing a lot of different things
- Making types support `Eq`
- More `From` impls
- A refactor with how the count math is done

etc

Ideally these would at least be split into individual commits
- We can track them in release notes more easily
- I can better see how all of the parts for a single change fit together

joshka replied:
Done

epage commented on Cargo.toml:
```diff
@@ -117,15 +117,19 @@ codecov = { repository = "clap-rs/clap-verbosity-flag" }
 [dependencies]
 clap = { version = "4.0.0", default-features = false, features = ["std", "derive"] }
 log = { version = "0.4.1", optional = true }
+serde = { version = "1", features = ["derive"], optional = true }
```
Is the `derive` worth it?  As this is a library meant for CLIs and they are already using `clap` with `derive` (which I would consider making optional) and the rest of the CLI is already likely using `serde`, this is probably fine.

Lurk commented:
I need that Serialise/Deserialize functionality and would like to continue work on it. 

As far as I understand, the way to go is to have three separate PRs: 

1. PR for how the count math is done
2. PR for `impl<L: LogLevel> From<Verbosity<L>> for VerbosityFilter` and `impl<L: LogLevel> From<VerbosityFilter> for Verbosity<L>` (this one will also include derive Eq for tests)
3. PR for feature-gated Serialize/Deserialize

Does it sound ok?

joshka commented:
> As far as I understand, the way to go is to have three separate PRs:

Likely just massaging the commits in this PR would do fine here. Ed has a high bar for small documented commits. Above force push is just a rebase onto the current main. Splitting in process now.

joshka commented on src/lib.rs:
```diff
@@ -512,4 +526,85 @@ mod test {
         }
     }
+
+    #[test]
+    fn from_verbosity_filter() {
+        for &filter in &[
+            VerbosityFilter::Off,
+            VerbosityFilter::Error,
+            VerbosityFilter::Warn,
+            VerbosityFilter::Info,
+            VerbosityFilter::Debug,
+            VerbosityFilter::Trace,
+        ] {
+            assert_eq!(Verbosity::<OffLevel>::from(filter).filter(), filter);
+            assert_eq!(Verbosity::<ErrorLevel>::from(filter).filter(), filter);
+            assert_eq!(Verbosity::<WarnLevel>::from(filter).filter(), filter);
+            assert_eq!(Verbosity::<InfoLevel>::from(filter).filter(), filter);
+            assert_eq!(Verbosity::<DebugLevel>::from(filter).filter(), filter);
+            assert_eq!(Verbosity::<TraceLevel>::from(filter).filter(), filter);
+        }
+    }
+}
```
Should be in previous commit
